cluster mempool: Implement changeset interface for mempool

sdaftuar opened the pull request:
part of cluster mempool: #30289 

It became clear while working on cluster mempool that it would be helpful for transaction validation if we could consider a full set of proposed changes to the mempool -- consisting of a set of transactions to add, and a set of transactions (ie conflicts) to simultaneously remove -- and perform calculations on what the mempool would look like if the proposed changes were to be applied.  Two specific examples of where we'd like to do this:

- Determining if ancestor/descendant/TRUC limits would be violated (in the future, cluster limits) if either a single transaction or a package of transactions were to be accepted
- Determining if an RBF would make the mempool "better", however that idea is defined, both in the single transaction and package of transaction cases

In preparation for cluster mempool, I have pulled this reworking of the mempool interface out of #28676 so it can be reviewed on its own.  I have not re-implemented ancestor/descendant limits to be run through the changeset, since with cluster mempool those limits will be going away, so this seems like wasted effort.  However, I have rebased #28676 on top of this branch so reviewers can see what the new mempool interface could look like in the cluster mempool setting.

There are some minor behavior changes here, which I believe are inconsequential:
- In the package validation setting, transactions would be added to the mempool before the `ConsensusScriptChecks()` are run. In theory, `ConsensusScriptChecks()` should always pass if the `PolicyScriptChecks()` have passed and it's just a belt-and-suspenders for us, but if somehow they were to diverge then there could be some small behavior change from adding transactions and then removing them, versus never adding them at all.
- The error reporting on `CheckConflictTopology()` has slightly changed due to no longer distinguishing between direct conflicts and indirect conflicts. I believe this should be entirely inconsequential because there shouldn't be a logical difference between those two ideas from the perspective of this function, but I did have to update some error strings in some tests.
- Because, in a package setting, RBFs now happen as part of the entire package being accepted, the logging has changed slightly because we do not know which transaction specifically evicted a given removed transaction. 
  -  Specifically, the "package hash" is now used to reference the set of transactions that are being accepted, rather than any single txid. The log message relating to package RBF that happen in the `TXPACKAGES` category has been updated as well to include the package hash, so that it's possible to see which specific set of transactions are being referenced by that package hash.
  - Relatedly, the tracepoint logging in the package rbf case has been updated as well to reference the package hash, rather than a transaction hash.

sdaftuar committed:
Introduce mempool changesets

Introduce the CTxMemPool::ChangeSet, a wrapper for creating (potential) new
mempool entries and removing conflicts.

sdaftuar committed:
Move changeset from workspace to subpackage

Removes a redundant check that mempool limits will not be violated during
package acceptance.

sdaftuar committed:
Clean up FinalizeSubpackage to avoid workspace-specific information

Also, use the "package hash" for logging replacements in the package rbf
setting.

sdaftuar committed:
Move LimitMempoolSize to take place outside FinalizeSubpackage

sdaftuar committed:
Apply mempool changeset transactions directly into the mempool

Rather than individually calling addUnchecked for each transaction added in a
changeset (after removing all the to-be-removed transactions), instead we can
take advantage of boost::multi_index's splicing features to extract and insert
entries directly from the staging multi_index into mapTx.

This has the immediate advantage of saving allocation overhead for mempool
entries which have already been allocated once. This also means that the memory
locations of mempool entries will not change when transactions go from staging
to the main mempool.

Additionally, eliminate addUnchecked and require all new transactions to enter
the mempool via a CTxMemPoolChangeSet.

sdaftuar committed:
Enforce that there is only one changeset at a time

sdaftuar committed:
Duplicate transactions are not permitted within a changeset

sdaftuar committed:
Don't distinguish between direct conflicts and all conflicts when doing cluster-size-2-rbf checks

sdaftuar committed:
Move CalculateChunksForRBF() to the mempool changeset

sdaftuar committed:
Move prioritisation into changeset

sdaftuar committed:
Public mempool removal methods Assume() no changeset is outstanding

While a changeset is outstanding, removing transaction directly from the
mempool will invalidate the changeset state, so this is not permitted.

sdaftuar committed:
Ensure that we don't add duplicate transactions in rbf fuzz tests

glozow commented on src/validation.cpp:
```diff
@@ -1295,36 +1295,41 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
-                workspaces[0].m_tx_handle->GetFee(),
-                workspaces[0].m_tx_handle->GetTxSize());
+                it->GetTxSize());
+        FeeFrac feerate = m_subpackage.m_changeset->GetAggregateFeeRate();
```
You could use `m_subpackage.m_total_modified_fees, m_subpackage.m_total_vsize` here? Unless you feel that `CTxMemPoolChangeSet::GetAggregateFeeRate` will be useful in the future.

sdaftuar replied:
Thanks that is much better; the only reason I had introduced `GetAggregateFeeRate()` was for access in this one place, so now that is gone.

glozow reviewed (commented):
Concept ACK, did a first pass

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -1295,36 +1295,41 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
-                workspaces[0].m_tx_handle->GetTxSize());
+                it->GetTxSize());
+        FeeFrac feerate = m_subpackage.m_changeset->GetAggregateFeeRate();
+        if (m_subpackage.m_changeset->GetTxCount() == 1) {
+            const CTransaction& tx = m_subpackage.m_changeset->GetTx(0);
+            LogDebug(BCLog::MEMPOOL, "New tx %s (wtxid=%s, fees=%s, vsize=%s)\n",
+                    tx.GetHash().ToString(),
+                    tx.GetWitnessHash().ToString(),
+                    feerate.fee, feerate.size);
+        } else {
+            LogDebug(BCLog::MEMPOOL, "New package with %lu txs, fees=%s, vsize=%s\n",
+                    m_subpackage.m_changeset->GetTxCount(),
+                    feerate.fee,
+                    feerate.size);
+        }
```
Should this be just logged once; i.e after replacing the transactions.

sdaftuar replied:
I tried to make this code analogous to the old code, which repeated information for each transaction.  I'm open to changing the logging but I am always a little afraid of breaking user's scripts or habits if people are used to grepping for a transaction ID and seeing more information? (Though I guess this PR already is breaking the logging for package RBF... so maybe not a big concern.)

ismaelsadeeq replied:
Yeah will break existing grepping scripts, but I think we will need to do this anyway to improve the logging for package replacements. 

I think we should log the new package or new transaction replacement only once after logging all the replaced transactions:

1. First, log all replaced transactions with their ID, witness, fee, and size.
2. Then:
   - For a single transaction replacement, log its ID, witness, fee, and size.
   - For a package replacement, log that a new package has been received, along with the package's number of transactions, fees, and vsize. Then, log each transaction ID and witness in the package.

This approach preserves the previous logging behavior for individual transactions.
Currently, in this PR, for a single transaction, the logs show the following:

```
2024-10-23T10:26:03.802679Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 6ae77ff52efae84e0cca125e4b5acfe2f3415298f4f98e850b40f89a2aa5990f (wtxid=57f1edd04c15f6532cdddf4a7c8393f5c5e5584f00003855df8c490053027ac8, fees=10000000, vsize=104).
2024-10-23T10:26:03.802686Z [httpworker.3] [src/validation.cpp:1317] [FinalizeSubpackage] [mempool] New tx e089a6391ae23d41baeed46fb1d8ac04cd0db73e82b4b330275f7d7cafcda33c (wtxid=db89e399619446928fb74a41a5207164beb0fd99972ba147b806a6a24f140e13, fees=490000000, vsize=104)
```

For a package, we see duplicate removal logs, which makes it seem like multiple packages are being added for each replacement:

```
2024-10-23T10:30:28.361089Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 05ad4c0db1a2fe169acd15bf5d8d1f4fa0ff834b5e8259baab4219f3d20b8991 (wtxid=39676091a39b5f77ee0865c2c52ac4bb8378ff7283fc1668738155ec247c3868, fees=10000, vsize=104).
2024-10-23T10:30:28.361110Z [httpworker.3] [src/validation.cpp:1322] [FinalizeSubpackage] [mempool] New package with 2 txs, fees=50000, vsize=208
2024-10-23T10:30:28.361127Z [httpworker.3] [src/validation.cpp:1310] [FinalizeSubpackage] [mempool] replacing mempool tx 992204e731c78b7afc6e4b17e14adf0f6b93d472fddc4ba184dd86e1899a83d1 (wtxid=05d9f03194089441530d7503a12b7f36086a2c41deac0960e238e426ed59e5f4, fees=10000, vsize=104).
2024-10-23T10:30:28.361139Z [httpworker.3] [src/validation.cpp:1322] [FinalizeSubpackage] [mempool] New package with 2 txs, fees=50000, vsize=208
```

With the idea above, the logs for individual transactions will remain the same, but for packages, the logs will look like this:

```
2024-10-23T11:46:49.459665Z [httpworker.2] [src/validation.cpp:1312] [operator()] [mempool] Replacing mempool tx 05ad4c0db1a2fe169acd15bf5d8d1f4fa0ff834b5e8259baab4219f3d20b8991 (wtxid=39676091a39b5f77ee0865c2c52ac4bb8378ff7283fc1668738155ec247c3868, fees=10000, vsize=104).
2024-10-23T11:46:49.459683Z [httpworker.2] [src/validation.cpp:1312] [operator()] [mempool] Replacing mempool tx 992204e731c78b7afc6e4b17e14adf0f6b93d472fddc4ba184dd86e1899a83d1 (wtxid=05d9f03194089441530d7503a12b7f36086a2c41deac0960e238e426ed59e5f4, fees=10000, vsize=104).
2024-10-23T11:46:49.459696Z [httpworker.2] [src/validation.cpp:1342] [FinalizeSubpackage] [mempool] New package (txs=2, fees=50000, vsize=208)
2024-10-23T11:46:49.459709Z [httpworker.2] [src/validation.cpp:1330] [operator()] [mempool] tx acd7ea806cd261b6c8cdb6d15210dad53daabbe68d5cb99f7df31646a769b69d (wtxid=8a867c2e5aea0d89831d3fe6831535e8cf8ce662ba36b6331fa072babc06f08c)
2024-10-23T11:46:49.459721Z [httpworker.2] [src/validation.cpp:1330] [operator()] [mempool] tx 28deb9d38a4490c3f542f0fd173805eb6c2c8309aa12d5e1669fc71a453d23fe (wtxid=8f8a74f38ed9deada97ef37bb5138b46d16f448da7585e33fc861b4142f764d5)
```

This will make the logs much easier to parse with a script. 
See naive diff implementing this change suggestion https://github.com/ismaelsadeeq/bitcoin/commit/555b05680d8e782fbe0d7b4807d2fce23f512b44

0xB10C replied:
Thanks for the ping, sorry for the late response. I agree that a "package hash" on it's own might not be useful for someone interested in "which transactions were replaced by which other transactions". On the other hand, the tracepoints are on purpose only semi-stable as they expose internals. If the internals change, the tracepoints need to change too.

 > which (as of this PR) would now report the package hash along with the transaction ids that are in it?

That could be useful, yes. However, constructing and passing arrays/lists to tracepoints is probably not something we want / can  do (at least not without https://github.com/bitcoin/bitcoin/pull/26593).

One solution could be to update the current tracepoint to a `TRACE8` and have a bool argument indicating if `tx_or_package_hash` is a tx or package hash with `m_subpackage.m_changeset->GetTxCount() == 1`. This wouldn't allow to get data on "which transactions were replaced by which other transactions", but consumers would know that the hash is either referring to a tx or package. While I'm currently using the tracepoint to monitor count, size and fees in replacements, I don't use the txids at the moment. I used them a while ago for [fullrbf.mempool.observer](https://web.archive.org/web/20230213002456/https://fullrbf.mempool.observer/). 

also, CC @virtu who initially worked on these tracepoints. Maybe you have another idea.

0xB10C replied:
> One solution could be to update the current tracepoint to a TRACE8 and have a bool argument indicating if tx_or_package_hash is a tx or package hash with m_subpackage.m_changeset->GetTxCount() == 1. 

This could look like e.g. https://github.com/0xB10C/bitcoin/commit/2eb1410e774636040d19d3baf5e22b2e4a8fbfd2

sdaftuar replied:
Thanks! I've included your commit (and will mark this as resolved).

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -1295,36 +1295,41 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
-                hash.data(),
-                workspaces[0].m_tx_handle->GetTxSize(),
-                workspaces[0].m_tx_handle->GetFee()
+                m_subpackage.m_changeset->GetTx(0).GetHash().data(),
```
Why are we only tracing the first tx in the package

sdaftuar replied:
Not sure if the tracing API is well-specified for package based replacement.  I'm open to doing something different than what I propose here though, suggestions?

ismaelsadeeq replied:
Its not specified, I think we have to specify tracing api for packages;

Should be something like

#### Tracepoint `mempool:replaced_by_package`

Is called when a transaction in the node's mempool is getting replaced by a package.
Passes information about the replaced transaction and  the replacement package.

Arguments passed:
1. Replaced transaction ID (hash) as `pointer to unsigned chars` (i.e. 32 bytes in little-endian)
2. Replaced transaction virtual size as `int32`
3. Replaced transaction fee as `int64`
4. Replaced transaction mempool entry time (epoch) as `uint64`
5. Replacement package virtual size as `int32`
6. Replacement package fee as `int64`

sdaftuar replied:
Tracking this issue above (https://github.com/bitcoin/bitcoin/pull/31122#discussion_r1814692827), so marking this conversation as resolved.

ismaelsadeeq reviewed (commented):
Concept ACK

The changes are straightforward to understand, and the large diff is primarily due to changes in the error message and the removal of `addUnchecked`.

sdaftuar commented on src/txmempool.h:
```diff
@@ -847,7 +847,14 @@ class CTxMemPool
-    std::unique_ptr<CTxMemPoolChangeSet> GetChangeSet() EXCLUSIVE_LOCKS_REQUIRED(cs) { return std::make_unique<CTxMemPoolChangeSet>(this); }
+    std::unique_ptr<CTxMemPoolChangeSet> GetChangeSet() EXCLUSIVE_LOCKS_REQUIRED(cs) {
+        Assume(m_changeset == nullptr);
+        auto ret = std::make_unique<CTxMemPoolChangeSet>(this);
+        m_changeset = ret.get();
+        return ret;
+    }
+
+    CTxMemPoolChangeSet* m_changeset GUARDED_BY(cs){nullptr};
```
This only exists for sanity checking that we don't have more than one changeset at a time.  We could just as easily replace this with a `bool`; I made it a pointer mostly for debugging purposes, if the `Assume()` were ever hit.

The idea behind the `unique_ptr` is that I wanted all the state in the ChangeSet to magically go away whenever the object goes out of scope, so that I didn't have to explicitly clean up state everywhere that a transaction might fail validation (which seems like a lot of places).  In this PR, there's not a ton of state, but I expect that when we implement cluster mempool that we'll be creating similar staging data structures within the `TxGraph` layer that will exist one level below the mempool (for computing clusters and feerate diagrams for candidate transactions), and I thought this mechanism of ending a staging when the changeset goes out of scope would be an easy way to manage things at that layer as well.  So in the future you could imagine that in the `CTxMemPoolChangeSet` constructor, we might initiate a staging in the underlying `TxGraph`; in the mempool `Apply()` function we might also apply the staging to the `TxGraph`; and in the `CTxMemPoolChangeSet` destructor we might abort a staging in the underlying `TxGraph`.

The paradigm I'm looking for is that you can have at most one changeset at a time, and when there's a changeset outstanding that serves as a lock on the mempool so that you can't otherwise add/remove things (except via the changeset) until the changes are either applied or aborted, and the changeset is the destroyed.  However if there's a better way to implement that than what I tried here, I'm certainly open to suggestions.

sdaftuar replied:
Done.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +816,40 @@ class CTxMemPool
+        using TxHandle = CTxMemPool::txiter;
+
+        TxHandle AddTx(const CTransactionRef& tx, const CAmount fee, int64_t time, unsigned int entry_height, uint64_t entry_sequence, bool spends_coinbase, int64_t sigops_cost, LockPoints lp);
+        void RemoveTx(CTxMemPool::txiter it) { m_all_conflicts.insert(it); }
```
Agreed, done.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +816,40 @@ class CTxMemPool
+        std::vector<CTxMemPool::txiter> m_entry_vec; // track the added transactions' insertion order
+        // map from the stage_tx index to the ancestors for the transaction
+        std::map<CTxMemPool::txiter, CTxMemPool::setEntries, CompareIteratorByHash> m_ancestors;
+        CTxMemPool::setEntries m_all_conflicts;
```
Agreed, done.

sdaftuar commented:
Thanks @glozow and @ismaelsadeeq for the review; I've incorporated some of the comments.  If you have suggestions for how the logging can be improved in the package RBF case (and how the USDT tracing messages can be improved, similarly) please let me know your thoughts.

Also, I wanted to flag that the locking is a bit of a mess in this branch.  Initially, I tried to add lock annotations to the `CTxMemPoolChangeSet` functions to guarantee that the mempool itself was locked whenever any of those functions were invoked.  I had mixed results with the compiler understanding those annotations, and ran into a roadblock at some point with a unit test which was taking the mempool lock and should have satisfied the annotations, but the compiler wasn't recognizing the lock as being the same one that needed to be held.  I assume the complication is mostly that the changeset is a different object than the mempool, so one thing I could try is moving all the changeset methods to instead be methods of the mempool itself (which take a changeset as an argument, potentially) -- but that also seems like a little bit of a mess to me, so wasn't sure if that was worth doing.

sdaftuar committed:
Add package hash to package-rbf log message

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -104,10 +104,18 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
-
     // Keep track of the total vsize of CTxMemPoolEntry's being added to the mempool to avoid overflow
     // Add replacement_vsize since this is added to new diagram during RBF check
+    std::optional<CMutableTransaction> replacement_tx = ConsumeDeserializable<CMutableTransaction>(fuzzed_data_provider, TX_WITH_WITNESS);
```
noting that this isn't actually required to get variable vsize transactions since `ConsumeTxMemPoolEntry` injects a fuzz-selected sigops value. Simpler if you just build a minimal tx then allow `ConsumeTxMemPoolEntry` to choose how big it is in vbytes?

sdaftuar replied:
I think I prefer letting the fuzzer pick it, so that I don't have to worry about whether there are other aspects of the transaction I'm overfitting (and thus getting worse coverage than if I let the fuzzer do all the work).

On the other hand, if there are certain transaction templates that we want to be sure are covered, then perhaps I can flip a bool and determine whether to use a pre-constructed transaction versus use one from the fuzzer.  

I can't think of what that might be though, so I'm inclined to leave this as-is unless you feel differently.

instagibbs replied:
I can't imagine it would greatly effect performance of the fuzz target as it's a single transaction, but it might matter more if more are constructed like in the rest of the harness.

instagibbs commented on src/policy/rbf.cpp:
```diff
@@ -184,14 +184,10 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
-                                                const CTxMemPool::setEntries& all_conflicts,
-                                                CAmount replacement_fees,
-                                                int64_t replacement_vsize)
+std::optional<std::pair<DiagramCheckError, std::string>> ImprovesFeerateDiagram(CTxMemPool::CTxMemPoolChangeSet& changeset)
```
at commit 1c33e133db5f876f2ba6eeb28877a5dcf3f6ac04 are we not applying priority values when computing RBFs, since they're only being applied during `Apply()` stage?

subsequent commit 2b5268ebac8daee0521ea82c2688e5e4fac885bc pulls it into changeset, meaning it can act on it?

Do we have coverage of package rbf with priority?

sdaftuar replied:
Yep you're right. I swapped the order of those two commits to avoid this problem, and I also added a test for package rbf with priority (which fails on https://github.com/bitcoin/bitcoin/commit/1c33e133db5f876f2ba6eeb28877a5dcf3f6ac04, but passes by the end of this branch).

instagibbs commented on src/validation.cpp:
```diff
@@ -1810,6 +1810,8 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
         AcceptSubPackage(txns_package_eval, args);
     PackageValidationState& package_state_final = multi_submission_result.m_state;
 
+    ClearSubPackageState();
```
This shouldn't be necessary, but shouldn't hurt either. `AcceptSubPackage` calls it already for each subpackage.

sdaftuar replied:
I think the issue is if we `quit_early`; I recall the fuzzer hitting the new `Assume()` calls via `LimitMempoolSize()` at the line just below, before I added this.

instagibbs replied:
Was the assert failed because of the changeset state, or something that could be hit in master?

`AcceptSubPackage` is called for each submission attempt of any subpackage, so I'm not sure why `quit_early` would be problematic.

instagibbs replied:
still curious about this if you figure out how it was hit

sdaftuar replied:
Now I think I may have been mistaken; I've commented out this line and started fuzzing again, and I cannot reproduce the crash. Also, I agree with your assessment after re-reading the code some more.

However, it would be nice if the SubPackageState were clearly destroyed when it goes out of scope from logic that uses it...  Even though it's seemingly not necessary to call cleanup here, it makes me a little uncomfortable to leave the code in a place where it takes some investigation to figure out that the changeset will be gone by the time LimitMempoolSize() is called.

I might just add a comment mentioning that this should have gotten cleaned up by `AcceptSubPackage()` but that I'm leaving the cleanup call for readability, given that the call to `LimitMempoolSize` requires that the changeset be gone.  Does that seem reasonable?

instagibbs commented on src/test/fuzz/rbf.cpp:
```diff
@@ -149,7 +154,9 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
             break;
         }
-        AddToMempool(pool, child_entry);
+        if (!pool.GetIter(child_entry.GetTx().GetHash())) {
```
belt-and-suspenders are fine, but it "shouldn't happen" since each tx chain is sourced from a new `g_outpoints`?

sdaftuar replied:
No I think this one is just "belt" -- if you look at lines 146-148 above, we're not always modifying the "child" transaction here, so sometimes it'll be a duplicate of something we already have.

instagibbs replied:
ah right

instagibbs reviewed (commented):
first pass through, concept ACK

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +816,40 @@ class CTxMemPool
+    private:
+        CTxMemPool* m_pool;
+        CTxMemPool::indexed_transaction_set m_to_add;
+        std::vector<CTxMemPool::txiter> m_entry_vec; // track the added transactions' insertion order
```
My intention was to allow calling `StageAddition` and `StageRemoval` in any order.  I can add comments to that effect.

sdaftuar replied:
Added docs in 7bc0f1f48c079b019f1d7a3fef00ae711f25d624

sdaftuar replied:
Update: I realized that at least for now, there is a requirement that any added transactions are added in a topological order, with parents coming before children -- this is so that we can add things to the mempool's multi_index at `Apply()` time without creating any missing dependencies.  But removals can happen in any order and can be interleaved with the additions.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1381,7 +1381,13 @@ void CTxMemPool::CTxMemPoolChangeSet::Apply()
     for (size_t i=0; i<m_entry_vec.size(); ++i) {
         auto tx_entry = m_entry_vec[i];
-        m_pool->addUnchecked(*tx_entry);
+        if (i == 0 && m_ancestors.count(tx_entry)) {
```
Yes this matches the old behavior, I believe.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1306,25 +1312,23 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
-    m_pool.RemoveStaged(m_subpackage.m_all_conflicts, false, MemPoolRemovalReason::REPLACED);
+    ws.m_changeset->Apply();
     // Don't attempt to process the same conflicts repeatedly during subpackage evaluation:
     // they no longer exist on subsequent calls to Finalize() post-RemoveStaged
```
Fixed.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1381,7 +1381,13 @@ void CTxMemPool::CTxMemPoolChangeSet::Apply()
     for (size_t i=0; i<m_entry_vec.size(); ++i) {
         auto tx_entry = m_entry_vec[i];
-        m_pool->addUnchecked(*tx_entry);
+        if (i == 0 && m_ancestors.count(tx_entry)) {
```
I think this is done?

sdaftuar commented on src/validation.cpp:
```diff
@@ -1467,11 +1471,16 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+    FinalizeSubpackage(args);
+
+    // Limit the mempool, if appropriate.
+    if (!args.m_bypass_limits) {
```
Agreed, I have restored this argument where `LimitMempoolSize()` is invoked in `AcceptSingleTransaction()`, and I pulled this change out into its own commit so that it's easier to review and less tangled with the other changes.

More thoughts on this [here](https://github.com/bitcoin/bitcoin/pull/28251#issuecomment-2433501732).  It would be nice to have a test that packages which involve a parent transaction that would be just at the bottom of the mempool without their children are able to make it in with a child, if the resulting package would score higher in the mempool than other transactions.

instagibbs replied:
I'm working on a test case for master for that case

sdaftuar replied:
I think this is all good now (your test in #31152 passes for me locally), so marking this as resolved.

glozow replied:
nice catch

sdaftuar committed:
test: Add unit test coverage of package rbf + prioritisetransaction

sdaftuar committed:
Remove m_all_conflicts from SubPackageState

sdaftuar commented on src/validation.cpp:
```diff
@@ -1312,20 +1315,21 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
     }
-    ws.m_changeset->Apply();
+    m_subpackage.m_changeset->Apply();
+    m_subpackage.m_changeset.reset();
```
After looking at this a little more, I don't think we need `m_all_conflicts` to live in `SubPackageState` anymore -- `FinalizeSubpackage` can grab this information from the changeset and then the other two places are just local uses within a function. 

Pushed a commit to get rid of this extra state.

sdaftuar committed:
doc: add comments for CTxMemPool::ChangeSet

instagibbs commented on src/txmempool.h:
```diff
@@ -816,6 +790,102 @@ class CTxMemPool
+    /*
+     * CTxMemPoolChangeSet:
+     *
+     * This class is used for all mempool additions and associated removals (eg due to rbf).
```
aside from reorgs?

sdaftuar replied:
Reorgs are split: additions happen via the changesets (since we just use ATMP), while removals happen in their own special funky way.

I can update the comment.

sdaftuar replied:
This is updated -- does this seem sufficient now, or should I add more documentation about how reorgs work?

instagibbs replied:
text is better thanks

instagibbs commented on src/txmempool.h:
```diff
@@ -816,6 +790,102 @@ class CTxMemPool
+     *
+     * Apply() will apply the removals and additions that are staged into the* mempool.
+     *
+     * Only one changeset may exist at a time. While a changeset is outstanding, no removals
```
when would removals or additions happen outside of a changeset?

sdaftuar replied:
Additions never happen outside a changeset.  RBF-related removals always go through a changeset as well.

Removals that happen because a block is found, or due to a reorg, or due to mempool expiry or mempool limiting do not go through the changeset.

instagibbs commented on src/txmempool.h:
```diff
@@ -816,6 +790,102 @@ class CTxMemPool
+     * CalculateMemPoolAncestors() calculates the in-mempool (not including
+     * what is in the change set itself) ancestors of a given transacion.
+     *
+     * Apply() will apply the removals and additions that are staged into the* mempool.
```
?
```suggestion
     * Apply() will apply the removals and additions that are staged into the mempool.
```

sdaftuar replied:
Fixed, thanks.

instagibbs commented on src/txmempool.h:
```diff
@@ -816,6 +816,40 @@ class CTxMemPool
+        CTxMemPool::indexed_transaction_set m_to_add;
+        std::vector<CTxMemPool::txiter> m_entry_vec; // track the added transactions' insertion order
+        // map from the m_to_add index to the ancestors for the transaction
+        std::map<CTxMemPool::txiter, CTxMemPool::setEntries, CompareIteratorByHash> m_ancestors;
```
logically seems to work, probably not worth nitting

sdaftuar replied:
I think this is resolved now with the rewrite of `ChangeSet::CalculateMemPoolAncestors()` -- let me know if you agree.

instagibbs replied:
this is fine, feel free to resolve it

instagibbs replied:
ok so I think I was a bit wrong, these values were being used and would "blind" us potentially from catching the "BUG! Mempool ancestors or descendants were underestimated" error.

instagibbs commented on src/validation.cpp:
```diff
@@ -1364,36 +1373,18 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
+    }
+    if (!all_submitted) {
+        // This code should be unreachable; it's here as belt-and-suspenders
+        // to try to ensure we have no consensus-invalid transactions in the
```
we're still applying on failure, so it's not really a belt and suspenders for that anymore is it?

sdaftuar replied:
Not sure I follow -- if somehow the PolicyScriptChecks() passed for something that failed ConsensusChecks(), then this logic would trigger and we would then remove the transactions from the mempool.  It's slightly different than the old behavior of never allowing it to enter in the first place, but given that we're holding the mempool locks the whole time it is not really directly observable?

instagibbs replied:
Misunderstood the code at time of writing this comment, mark as resolved

instagibbs commented on src/validation.cpp:
```diff
@@ -1100,13 +1100,15 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
                              strprintf("insufficient fee%s", ws.m_sibling_eviction ? " (including sibling eviction)" : ""), *err_string);
     }
 
+    CTxMemPool::setEntries m_all_conflicts;
```
```suggestion
    CTxMemPool::setEntries all_conflicts;
```

sdaftuar replied:
Fixed, thanks.

instagibbs commented on src/validation.cpp:
```diff
@@ -1178,13 +1185,16 @@ bool MemPoolAccept::PackageMempoolChecks(const std::vector<CTransactionRef>& txn
 
     // Don't consider replacements that would cause us to remove a large number of mempool entries.
     // This limit is not increased in a package RBF. Use the aggregate number of transactions.
+    CTxMemPool::setEntries m_all_conflicts;
```
```suggestion
    CTxMemPool::setEntries all_conflicts;
```

sdaftuar replied:
Fixed, thanks.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,102 @@ class CTxMemPool
+     * Callers can invoke StageAddition()/StageRemoval() of transactions in any order.
+     *
+     * CalculateChunksForRBF() can be used to calculate the feerate diagram of the proposed
+     * set of new transactions and comare with the existing mempool.
```
Fixed, thanks.

sdaftuar commented:
Rebased (to pick up the test introduced in #31152) and addressed latest feedback.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,111 @@ class CTxMemPool
+     * outstanding, no removals or additions may be made directly to the
+     * mempool.
+     */
+    class CTxMemPoolChangeSet {
```
Done.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,111 @@ class CTxMemPool
+        util::Result<CTxMemPool::setEntries> CalculateMemPoolAncestors(TxHandle tx, const Limits& limits) {
+            LOCK(m_pool->cs);
+            auto ret = m_pool->CalculateMemPoolAncestors(*tx, limits);
+            if (ret) m_ancestors.insert({tx, *ret});
```
No longer relevant after rewriting this function.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,111 @@ class CTxMemPool
+
+        const CTxMemPool::setEntries& GetRemovals() const { return m_to_remove; }
+
+        util::Result<CTxMemPool::setEntries> CalculateMemPoolAncestors(TxHandle tx, const Limits& limits) {
```
I reworked your suggestion to capture the caching idea, and now invoke this function in `Apply`.  However, there's still an ugly special case in `Apply()` in that we can only used a cached ancestor set calculation for the first transaction in a package (because the calculation is no longer correct once transactions start getting added to the mempool).

Note that in the cluster mempool implementation, this ugliness in `Apply()` goes away, because we will no longer need to have ancestor sets calculated in order to add a transaction to the mempool.  

Furthermore, I believe that we may be able to completely eliminate the need for this function altogether, post-cluster-mempool.  I should go back and try to rework #28676 to demonstrate that this is the case, but my recollection is that the only reason we will need to calculate ancestors sets in the future is for an RBF sanity check (namely that we're not replacing an ancestor of any to-be-added transaction) which we can reimplement without these ancestor calculations.

sdaftuar replied:
Looking into whether I can avoid exposing this function even before cluster mempool lands; will consider it for a followup PR.

instagibbs commented on src/txmempool.h:
```diff
@@ -865,9 +856,25 @@ class CTxMemPool
+    friend class CTxMemPool::ChangeSet;
+
+private:
+    // addNewTransaction must updated state for all ancestors of a given transaction,
```
while we're moving the comment, can we fix the grammar: "must updated"?

sdaftuar replied:
Fixed.

instagibbs commented on src/txmempool.h:
```diff
@@ -865,9 +856,25 @@ class CTxMemPool
+    // Note that addNewTransaction is ONLY called from ATMP outside of tests
+    // and any other callers may break wallet's in-mempool tracking (due to
+    // lack of CValidationInterface::TransactionAddedToMempool callbacks).
+    void Apply(CTxMemPool::ChangeSet* changeset) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
comment above this  isn't for `Apply`.

sdaftuar replied:
Moved.

instagibbs commented on src/txmempool.h:
```diff
@@ -865,9 +856,25 @@ class CTxMemPool
+    // to track size/count of descendant transactions.  First version of
+    // addNewTransaction can be used to have it call CalculateMemPoolAncestors(), and
+    // then invoke the second version.
+    // Note that addNewTransaction is ONLY called from ATMP outside of tests
```
it's not called from ATMP or tests anymore, thankfully it's private now

sdaftuar replied:
Reworded.

instagibbs commented on src/txmempool.h:
```diff
@@ -260,7 +260,7 @@ struct TxMempoolInfo
  * children (because any such children would be an orphan).  So in
- * addUnchecked(), we:
+ * addNewTransaction(), we:
  * - update a new entry's setMemPoolParents to include all in-mempool parents
```
not your fault but: I don't think `setMemPoolParents` exists in master even

I feel like these paragraphs are a minefield of old information

sdaftuar replied:
Tried to remove all instances of setMemPoolParents and setMemPoolChildren from code comments.

instagibbs commented on src/txmempool.cpp:
```diff
@@ -603,6 +603,7 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
 {
```
c8b896067d42001f2a47cd837c0aed07bb39f329 commit message could probably use a little massaging since removals can happen in a changeset? I'm assuming these are all block/trimming related

sdaftuar replied:
This was a good suggestion, thanks -- with the introduction of changesets, we can make `RemoveStaged()` a private method now, and so I updated the commit message to explain that the public removal methods may not be invoked while a changeset is outstanding.

instagibbs commented on src/test/rbf_tests.cpp:
```diff
@@ -374,6 +374,10 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
 
+    // conflicting transactions
+    const auto tx1_conflict = make_tx(/*inputs=*/ {m_coinbase_txns[0], m_coinbase_txns[2]}, /*output_values=*/ {10 * COIN});
+    auto entry1_conflict = entry.FromTx(tx1_conflict);
```
5e447a3059224ac97ec7cf5681dc3ca492770968 unused?

sdaftuar replied:
Gone thanks

instagibbs commented on src/validation.cpp:
```diff
@@ -740,6 +739,8 @@ class MemPoolAccept
         CTxMemPool::setEntries m_all_conflicts;
         /** Mempool transactions that were replaced. */
         std::list<CTransactionRef> m_replaced_transactions;
+        /* Changeset representing adding a transaction and removing its conflicts. */
```
circa 27c2abffbd9c5a83fd2becf6d7b70c7b30031f76 it's applying multiple transactions
```Suggestion
        /* Changeset representing adding transactions and removing their conflicts. */
```

sdaftuar replied:
Done

instagibbs commented on src/txmempool.cpp:
```diff
@@ -1309,11 +1309,15 @@ std::optional<std::string> CTxMemPool::CheckConflictTopology(const setEntries& d
+    FeeFrac replacement_feerate{0, 0};
+    for (auto it : m_entry_vec) {
+        replacement_feerate += {it->GetModifiedFee(), it->GetTxSize()};
+    }
```
after this loop, sanity check replacement_feerate isn't size 0?

instagibbs replied:
Ok, given new interface seems impossible to hit. Added another test case to `ComapreChunks` to just make sure it was well-defined:

```
diff --git a/src/test/rbf_tests.cpp b/src/test/rbf_tests.cpp
index c6d06c937c..d610661878 100644
--- a/src/test/rbf_tests.cpp
+++ b/src/test/rbf_tests.cpp
@@ -631,20 +631,32 @@ BOOST_AUTO_TEST_CASE(feerate_chunks_utilities)
 {
     // Sanity check the correctness of the feerate chunks comparison.
 
     // A strictly better case.
     std::vector<FeeFrac> old_chunks{{{950, 300}, {100, 100}}};
     std::vector<FeeFrac> new_chunks{{{1000, 300}, {50, 100}}};
 
     BOOST_CHECK(std::is_lt(CompareChunks(old_chunks, new_chunks)));
     BOOST_CHECK(std::is_gt(CompareChunks(new_chunks, old_chunks)));
 
+    // Empty comparisons
+    old_chunks = {{0, 0}};
+    new_chunks = {{1000, 300}, {0, 100}};
+
+    BOOST_CHECK(std::is_lt(CompareChunks(old_chunks, new_chunks)));
+    BOOST_CHECK(std::is_gt(CompareChunks(new_chunks, old_chunks)));
+
+    old_chunks = {{0, 0}};
+    new_chunks = {{0, 0}};
+
+    BOOST_CHECK(std::is_eq(CompareChunks(old_chunks, new_chunks)));
+
     // Incomparable diagrams
     old_chunks = {{950, 300}, {100, 100}};
     new_chunks = {{1000, 300}, {0, 100}};
 
     BOOST_CHECK(CompareChunks(old_chunks, new_chunks) == std::partial_ordering::unordered);
     BOOST_CHECK(CompareChunks(new_chunks, old_chunks) == std::partial_ordering::unordered);
 
     // Strictly better but smaller size.
     old_chunks = {{950, 300}, {100, 100}};
     new_chunks = {{1100, 300}};
```

sdaftuar replied:
Maybe worth a separate PR to ensure the test coverage.

instagibbs reviewed (commented):
LGTM e2324779781b695024d3b17340001fa2da5d3c7f

with non-blocking nits

will do some testing

naumenkogs commented on src/validation.cpp:
```diff
@@ -682,7 +682,9 @@ class MemPoolAccept
 
     // Try to add the transaction to the mempool, removing any conflicts first.
-    void FinalizeSubpackage(const ATMPArgs& args, std::vector<Workspace>& ws) EXCLUSIVE_LOCKS_REQUIRED(cs_main, m_pool.cs);
+    // Returns true if the transaction is in the mempool after any size
```
in a8d0976de7edf43d3cfea8dff3afc923580d8f84

sdaftuar replied:
No idea!  Gone now.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1366,7 +1370,7 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
         // last calculation done in PreChecks, since package ancestors have already been submitted.
         {
-            auto ancestors{m_pool.CalculateMemPoolAncestors(*ws.m_entry, m_pool.m_opts.limits)};
+            auto ancestors{ws.m_changeset->CalculateMemPoolAncestors(ws.m_tx_handle, m_pool.m_opts.limits)};
```
I'm very puzzled why none of the tests appear to fail on this commit.

sdaftuar replied:
Ah, it's because at this point, the `Apply()` function doesn't use the ancestors anyway.  Agreed that this is pretty ugly -- I'll try to fix.

sdaftuar replied:
Should be better now, I think -- now I'm introducing the ancestor caching inside of `CalculateMemPoolAncestors` into the same commit that introduces using the cached value in Apply().

instagibbs replied:
history is cleaner, thank you

sdaftuar committed:
Make RemoveStaged() private

sdaftuar commented on src/validation.cpp:
```diff
@@ -1365,36 +1370,18 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
-                                  strprintf("BUG! Adding to mempool failed: %s", ws.m_ptx->GetHash().ToString()));
-        }
+    }
+    if (!all_submitted) {
```
Added an `Assume()` for it.

instagibbs commented:
ACK 9512bd34a3d74d15c3ce1086a80a1edfa5ac3acf

via `git range-diff master e2324779781b695024d3b17340001fa2da5d3c7f 9512bd34a3d74d15c3ce1086a80a1edfa5ac3acf`

sdaftuar commented on src/txmempool.h:
```diff
@@ -844,6 +844,18 @@ class CTxMemPool
+        }
+
+        size_t GetTxCount() const { return m_entry_vec.size(); }
+        const CTransaction& GetTx(size_t index) const { return m_entry_vec.at(index)->GetTx(); }
```
Done.

ismaelsadeeq reviewed (commented):
ACK 9512bd34a3d74d15c3ce1086a80a1edfa5ac3acf
With some comments.

0xB10C committed:
tracing: pass if replaced by tx/pkg to tracepoint

The mempool:replaced tracepoint now reports either a txid or a
package hash (previously it always was a txid). To let users know
if a txid or package hash is passed, a boolean argument is added
the the tracepoint.

In the functional test, a ctypes.Structure class for MempoolReplaced
is introduced as Python warns the following when not explcitly
casting it to a ctype:

  Type: 'bool' not recognized. Please define the data with ctypes manually.

sdaftuar commented on src/test/txpackage_tests.cpp:
```diff
@@ -1078,7 +1078,25 @@ BOOST_AUTO_TEST_CASE(package_rbf_tests)
         BOOST_CHECK_EQUAL(m_node.mempool->size(), expected_pool_size);
-    }
 
+        // Finally, check that we can prioritise package1 and get it back into the mempool.
```
Done

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +816,37 @@ class CTxMemPool
+    private:
+        CTxMemPool* m_pool;
+        CTxMemPool::indexed_transaction_set m_to_add;
+        std::vector<CTxMemPool::txiter> m_entry_vec; // track the added transactions' insertion order
```
My thought was that we're storing `CTxMemPool::txiter` objects internally, but the `TxHandle` object could be replaced with some other wrapper if we wanted to hide the internals of the mempool further.  So I'm leaving this as-is for now.

ismaelsadeeq replied:
Ahh Makes sense

sdaftuar commented on src/validation.cpp:
```diff
@@ -1365,36 +1370,19 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
-        // last calculation done in PreChecks, since package ancestors have already been submitted.
-        {
-            auto ancestors{ws.m_changeset->CalculateMemPoolAncestors(ws.m_tx_handle, m_pool.m_opts.limits)};
-            if(!ancestors) {
```
Updated the commit message.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1331,17 +1327,7 @@ bool MemPoolAccept::FinalizeSubpackage(const ATMPArgs& args, std::vector<Workspa
-            return state.Invalid(TxValidationResult::TX_RECONSIDERABLE, "mempool full");
-    }
-    return true;
+    return;
```
Gone.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1290,34 +1290,48 @@ bool MemPoolAccept::ConsensusScriptChecks(const ATMPArgs& args, Workspace& ws)
-                hash.data(),
-                workspaces[0].m_tx_handle->GetTxSize(),
-                workspaces[0].m_tx_handle->GetFee()
+                tx_or_package_hash.data(),
```
I cherry-picked https://github.com/0xB10C/bitcoin/commit/2eb1410e774636040d19d3baf5e22b2e4a8fbfd2, which updates the documentation and resolves the issue discussed above: https://github.com/bitcoin/bitcoin/pull/31122#discussion_r1832803900

0xB10C commented:
> 🐙 This pull request conflicts with the target branch and [needs rebase](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes).

Note that in #26593 the tracepoint macros (e.g. `TRACE7`, `TRACE8`, ..) were renamed to just `TRACEPOINT` (without a number). When rebasing, there will be a few places where these will need to be renamed as you touched code close to it. Similarly, in the last commit ("tracing: pass if replaced by tx/pkg to tracepoint", currently 0ca1e05d8bcbc42892156c15f128a5f829e9e48e) the `TRACE8` won't work and will need to be just `TRACEPOINT`.

naumenkogs commented:
ACK 0ca1e05d8bcbc42892156c15f128a5f829e9e48e

Spent a fair amount of time staring at `ConsensusScriptChecks` behavior change mentioned in the original post. It seems alright. Handling various limits is probably the most complex part of this code change.

Otherwise the code seems clear and nothing is broken. Still hard for me to reason about high-level design decisions on cluster mempool here, but I learned a lot while reviewing this pr at least.

instagibbs reviewed (approved):
ACK 0ca1e05d8bcbc42892156c15f128a5f829e9e48e

primary change is the tracing commit, which seems to make sense though I am not an expert

via `git range-diff master 9512bd34a3d74d15c3ce1086a80a1edfa5ac3acf 0ca1e05d8bcbc42892156c15f128a5f829e9e48e`

sdaftuar commented:
Just rebased this on master -- @0xB10C thanks for the note on the tracepoint change.  @instagibbs I think I could rewrite the `CheckEphemeralSpends()` code to take advantage of the changesets for package validation (since we will now be maintaining a map of package transactions that we can query), but maybe that can be done in a separate PR?  Just left that code as-is for now.

instagibbs reviewed (approved):
ACK 5736d1ddacc4019101e7a5170dd25efbc63b622a

Picks up TRACEPOINT and ephemeral dust stuff.

via `git range-diff master 0ca1e05d8bcbc42892156c15f128a5f829e9e48e 5736d1ddacc4019101e7a5170dd25efbc63b622a`

naumenkogs commented:
@sdaftuar 

>EDIT: I hacked up the validation logic to trigger this currently unreachable block of code -- ie to make the ConsensusScriptChecks() fail in a package validation context -- and verified that logic to remove the newly added transactions seems to work. Would be great for others to verify this as well!

I had a little doubt that it will remove the transactions... Instead I was worried about evictions (i can't think of other side effects) of add/remove invalid stuff.  Doesn't this allow to clear the victims mempool for free?

sdaftuar commented:
> Instead I was worried about evictions (i can't think of other side effects) of add/remove invalid stuff. Doesn't this allow to clear the victims mempool for free?

Just to be clear, in case anyone has lost track: we're specifically talking about a codepath we believe is not possible to trigger, namely one where the `PolicyScriptChecks` pass yet the `ConsensusScriptChecks` fail for some transaction, and someone is able to submit such a transaction as part of a package to our node.

In this event, I actually don't think the mempool could be directly drained, because we don't call `LimitMempoolSize()` until after this condition is detected and the new transactions are removed.  However, I did observe that any RBF's that take place would not be fully reverted.

If it were simple to just re-add the just-removed transactions, I'd try to include that here as well, but it would take some additional work to topologically sort the removed transactions so they can be re-added.  Given that this logic is unreachable and already difficult-to-test as a result, I don't think it makes sense to add any significant complexity as a result, at least for now -- this may be simpler to think about post-cluster-mempool.

ismaelsadeeq reviewed (commented):
reACK 5736d1ddacc4019101e7a5170dd25efbc63b622a

naumenkogs commented:
ACK https://github.com/bitcoin/bitcoin/commit/5736d1ddacc4019101e7a5170dd25efbc63b622a

My concern above was related to LimitMempoolSize(), but indeed it's easy to see it doesn't happen between the two `m_changeset->Apply()` calls .

glozow reviewed (commented):
Still working on review. I've skimmed through the commits and reviewed the end state of `MemPoolAccept` so far.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,122 @@ class CTxMemPool
+    /*
+     * CTxMemPool::ChangeSet:
+     *
+     * This class is used for all mempool additions and associated removals (eg
```
I've discussed this question with @sipa a bit; our thought was that changesets are for evaluating changes that we might need to back out and not apply; removals due to a block being found or to make the mempool consistent after a reorg have to happen, so we don't bother with a changeset and just apply them directly.

If you have other thoughts on this let me know-- happy to reconsider if there's a good reason to structure removals differently.

glozow replied:
That makes sense, thanks. I wanted to clarify what was meant by "all" additions and removals.

glozow replied:
Sorry about this comment - just reread what I said and it makes no sense.

sdaftuar commented on src/txmempool.h:
```diff
@@ -816,6 +790,122 @@ class CTxMemPool
+            auto it = m_ancestors.find(tx);
+            if (it != m_ancestors.end()) return it->second;
+
+            // If not found, try to have the mempool calculate it, and cache
+            // for later.
+            LOCK(m_pool->cs);
+            auto ret{m_pool->CalculateMemPoolAncestors(*tx, limits)};
+            if (ret) m_ancestors.try_emplace(tx, *ret);
+            return ret;
```
I also have a goal of eliminating the need to invoke `CMPA` on a changeset anyway.  In #28676, I believe the only place that `CMPA` is used in a changeset is for determining whether an RBF might be replacing something it depends on.  I think it should be straightforward to have the changeset logic directly test for that condition and return failure if a dependency is being added to something that is slated for removal, which would then allow us to remove the method altogether.

glozow commented on src/txmempool.cpp:
```diff
@@ -1163,6 +1167,7 @@ void CTxMemPool::trackPackageRemoved(const CFeeRate& rate) {
 
 void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpendsRemaining) {
     AssertLockHeld(cs);
+    Assume(!m_have_changeset);
```
b447416fddcb8c8647391502cca3dbfd1552e02e is nice!

glozow commented on src/test/rbf_tests.cpp:
```diff
@@ -374,6 +374,9 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
     const auto tx2_size = entry2->GetTxSize();
 
+    // conflicting transactions
+    const auto tx1_conflict = make_tx(/*inputs=*/ {m_coinbase_txns[0], m_coinbase_txns[2]}, /*output_values=*/ {10 * COIN});
```
nit: these changes were perhaps meant for d7dc9fd2f7bc675256687b9c55fdbec9cc8ac781 instead of 284a1d33f1dcbc3b3404ea40a948ff6600239613?

glozow reviewed (commented):
ACK 5736d1ddacc

I really like how `ChangeSet` codifies the concept of an atomic changeset of additions and removals, and how `CalculateChunksForRBF` and `ImprovesFeerateDiagram` operate on changesets. It's like the mempool interface is moving towards it being more responsible for itself.
